tests/test_downloadermiddleware.py

```diff
@@ -206,3 +209,46 @@ def process_request(self, request, spider):
 
         self.assertIs(results[0], resp)
         self.assertFalse(download_func.called)
+
+
+class MiddlewareUsingCoro(ManagerTestCase):
+    """Middlewares using asyncio coroutines should work"""
+
+    def test_asyncdef(self):
+        resp = Response('http://example.com/index.html')
+
+        class CoroMiddleware:
+            async def process_request(self, request, spider):
+                await defer.succeed(42)
+                return resp
+
+        self.mwman._add_middleware(CoroMiddleware())
+        req = Request('http://example.com/index.html')
+        download_func = mock.MagicMock()
+        dfd = self.mwman.download(download_func, req, self.spider)
+        results = []
+        dfd.addBoth(results.append)
+        self._wait(dfd)
+
+        self.assertIs(results[0], resp)
+        self.assertFalse(download_func.called)
+
+    @mark.only_asyncio()
+    def test_asyncdef_asyncio(self):
+        resp = Response('http://example.com/index.html')
+
+        class CoroMiddleware:
+            async def process_request(self, request, spider):
+                await asyncio.sleep(0.1)
```

Comment:
Could you please also add a test where some result is awaited?

Reply:
Such as?
Looks like asyncio without 3rd-party modules supports async networking, queues and subprocesses, not sure if anything else. Or you can create a Task and manage it.
We would also want similar code in all other asyncio testcases?

Reply:
I'm not sure which library to use and what action to perform exactly, but we need to test that results work - getting something from DB or making a HTTP request can be a standard use case for async def methods. If it is not easy to set something up using stdlib, that's fine to have additional requirements (e.g. aiohttp). That said, it should be possible to implement something using subprocess or queue.

Reply:
I've added a Queue with sync put and async get, it's not very interesting though, let me know if we should use something heavier.

Reply:
Looks good!